### PR TITLE
Fix rmm build

### PR DESCRIPTION
**Pull request by trivialfis**
- Optionally switch to c++17
- Use rmm CMake target. Close https://github.com/dmlc/xgboost/issues/7873 .
- Workaround compiler errors.
- Fix GPUMetric inheritance.
- Run death tests even if it's built with RMM support.

### CODE REVIEW

**Comment by trivialfis** [comment]
```
[2022-06-05T13:20:06.708Z] Error: Internal Compiler Error (codegen): "there was an error in verifying the lgenfe output!"
```

File: cmake/Utils.cmake [hunk 1]
```diff
@@ -169,10 +169,17 @@ function(xgboost_set_cuda_flags target)
...
-  set_target_properties(${target} PROPERTIES
-    CUDA_STANDARD 14
-    CUDA_STANDARD_REQUIRED ON
-    CUDA_SEPARABLE_COMPILATION OFF)
+  if (PLUGIN_RMM)
+    set_target_properties(${target} PROPERTIES
+      CUDA_STANDARD 17
+      CUDA_STANDARD_REQUIRED ON
+      CUDA_SEPARABLE_COMPILATION OFF)
+  else (PLUGIN_RMM)
```
Comment (jakirkham) [inline cmake/Utils.cmake:177]: Should this be?

```suggestion
  else ()
```

File: src/data/data.cu [hunk 2]
```diff
@@ -130,12 +130,12 @@ void MetaInfo::SetInfoFromCUDA(Context const&, StringView key, Json array) {
   }
   // uint info
   if (key == "group") {
-    auto array_interface{ArrayInterface<1>(array)};
+    ArrayInterface<1> array_interface{array};
```
Comment (jakirkham) [inline src/data/data.cu:133]: Curious why `auto` needed to be changed or is this just to be more explicit?

**Comment by jakirkham** [comment]
Thanks Jiaming! 🙏

Had a couple questions below

**Review by jakirkham — COMMENTED** [review]
Thanks Jiaming! 🙏

Had a couple questions below

File: src/data/data.cu (hunk 2 above)
Comment (trivialfis) [reply src/data/data.cu:133]: Workaround for compiler error.

**Comment by trivialfis** [comment]
For some reason, the CI is failing for other PRs as well. New commits are just used for debugging.

**Comment by trivialfis** [comment]
Couldn't reproduce the error locally.

**Comment by trivialfis** [comment]
Couldn't reproduce with docker + sanitizer.

**Comment by trivialfis** [comment]
And it's working?

**Comment by trivialfis** [comment]
Fixed by https://github.com/dmlc/xgboost/pull/7973/commits/2e7f216a916ce42a4886dda90f20a100e480acab

**Review by RAMitchell — APPROVED** [review]

**Pull request merged**